src/Encoding.chpl

```diff
@@ -0,0 +1,254 @@
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc uint32Consume(ch: channel(false,iokind.little,false)): uint(32) throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:uint(32);
+  }
+  
+
+  proc int64Append(val: int(64), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc int64Consume(ch: channel(false,iokind.little,false)): int throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:int(64);
+  }
+
+  proc int32Append(val: int(32), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc int32Consume(ch: channel(false,iokind.little,false)): int(32) throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:int(32);
+  }
+
+  proc boolAppend(val: bool, ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = val:uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc boolConsume(ch: channel(false,iokind.little,false)): bool throws {
+    var (val, len) = unsignedVarintConsume(ch);
+    return val:bool;
+  }
+
+  proc sint64Append(val: int(64), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = (val << 1):uint ^ (val >> 63):uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc sint64Consume(ch: channel(false,iokind.little,false)): int(64) throws {
+    const (val, len) = unsignedVarintConsume(ch);
+    return (val >> 1):int(64) ^ (val):int(64) << 63 >> 63;
+  }
+
+  proc sint32Append(val: int(64), ch: channel(true,iokind.little,false)) throws {
+    const wireType = varint;
+    var uintVal = (val << 1):uint ^ (val >> 31):uint;
+    unsignedVarintAppend(uintVal, ch);
+  }
+
+  proc sint32Consume(ch: channel(false,iokind.little,false)): int(32) throws {
+    const (val, len) = unsignedVarintConsume(ch);
+    return (val >> 1):int(32) ^ (val):int(32) << 31 >> 31;
+  }
+
+  proc bytesAppend(val: bytes, ch: channel(true,iokind.little,false)) throws {
+    const wireType = lengthDelimited;
+    unsignedVarintAppend((val.size):uint, ch);
+    ch.write(val);
+  }
+
+  proc bytesConsume(ch: channel(false,iokind.little,false)): bytes throws {
+    const (byteLen, len) = unsignedVarintConsume(ch);
+    var s:bytes;
+    ch.readbytes(s, byteLen:int);
+    return s;
+  }
+
+  proc stringAppend(val: string, ch: channel(true,iokind.little,false)) throws {
+    bytesAppend(val.encode(), ch);
+  }
+
+  proc stringConsume(ch: channel(false,iokind.little,false)): string throws {
+    return bytesConsume(ch).decode();
+  }
+
+
+  proc fixed32Append(val: uint(32), ch: channel(true,iokind.little,false)) throws {
+    const wireType = fixed32Type;
+    for i in 0..24 by 8 {
+      var newByte = (val >> i):uint(8);
+      ch.writeBytes(newByte, 1);
```

Comment:
This can just be `ch.write(newByte)` but more generally this is just outputting the 32-bit value in big-endian format. Which makes me wonder if iokind.big would be a better setting for the channels? If iokind.big is the setting, then this would just be `ch.write(val)`.

This comment applies similarly to all of the fixed append and consume functions.

Reply:
> This comment applies similarly to all of the fixed append and consume functions.

I am also a bit confused regarding the encoding/decoding of fixed types. The protobuf-encoding guide states them to be encoded as little-endian, but I saw a similar kind of implementation in Go protbuf library by Google(https://github.com/protocolbuffers/protobuf-go/blob/master/encoding/protowire/wire.go#L373).

Reply:
> This can just be `ch.write(newByte)`

Ohh! I missed that, will change it :+1:

Reply:
> I am also a bit confused regarding the encoding/decoding of fixed types. The protobuf-encoding guide states them to be encoded as little-endian, but I saw a similar kind of implementation in Go protbuf library by Google(https://github.com/protocolbuffers/protobuf-go/blob/master/encoding/protowire/wire.go#L373).

That go implementation is writing it little endian. So is your loop here, I think I read it wrong before. Anyway you should be able to just write `ch.write(val)`. Give it a try?

Reply:
Yes, it works!